core/tee/tee_enc_fs_key_manager.c

```diff
@@ -0,0 +1,362 @@
+	}
+
+exit:
+	return res;
+}
+
+static TEE_Result do_file_encryption(struct enc_fs_file_header *hdr,
+			uint8_t *fek, int fek_len,
+			uint8_t *data_in, size_t in_size,
+			uint8_t *data_out, size_t *out_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t *ctx = NULL;
+	size_t ctx_size;
+	uint8_t *dst_data = data_out + sizeof(struct enc_fs_file_header);
+	size_t dst_len = *out_size - sizeof(struct enc_fs_file_header);
+	size_t tag_len = ENC_FS_MAX_TAG_LEN;
+
+	res = crypto_ops.authenc.get_ctx_size(ENC_FS_AUTH_ENC_ALG, &ctx_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	ctx = malloc(ctx_size);
+	if (ctx == NULL)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = crypto_ops.authenc.init(ctx, ENC_FS_AUTH_ENC_ALG,
+			TEE_MODE_ENCRYPT, fek, fek_len, hdr->aad.iv,
+			ENC_FS_IV_LEN, ENC_FS_MAX_TAG_LEN,
+			sizeof(struct aad), in_size);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.authenc.update_aad(ctx, ENC_FS_AUTH_ENC_ALG,
+			TEE_MODE_ENCRYPT, (uint8_t *)&hdr->aad,
+			sizeof(struct aad));
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.authenc.enc_final(ctx, ENC_FS_AUTH_ENC_ALG,
+			data_in, in_size, dst_data, &dst_len,
+			hdr->tag, &tag_len);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	crypto_ops.authenc.final(ctx, ENC_FS_AUTH_ENC_ALG);
+
+	memcpy(data_out, hdr, sizeof(*hdr));
+	*out_size = sizeof(*hdr) + dst_len;
+
+exit:
+	free(ctx);
+	return res;
+}
+
+static TEE_Result do_file_decryption(struct enc_fs_file_header *hdr,
+			uint8_t *fek, int fek_len,
+			uint8_t *ciphertext, size_t ciphertext_size,
+			uint8_t *plaintext, size_t *plaintext_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	uint8_t *ctx = NULL;
+	size_t ctx_size;
+	size_t tag_len = ENC_FS_MAX_TAG_LEN;
+
+	res = crypto_ops.authenc.get_ctx_size(ENC_FS_AUTH_ENC_ALG, &ctx_size);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	ctx = malloc(ctx_size);
+	if (ctx == NULL)
+		return TEE_ERROR_OUT_OF_MEMORY;
+
+	res = crypto_ops.authenc.init(ctx, ENC_FS_AUTH_ENC_ALG,
+			TEE_MODE_DECRYPT, fek, fek_len,
+			hdr->aad.iv, ENC_FS_IV_LEN, ENC_FS_MAX_TAG_LEN,
+			sizeof(struct aad), ciphertext_size);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.authenc.update_aad(ctx, ENC_FS_AUTH_ENC_ALG,
+			TEE_MODE_DECRYPT, (uint8_t *)&hdr->aad,
+			sizeof(struct aad));
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	res = crypto_ops.authenc.dec_final(ctx, ENC_FS_AUTH_ENC_ALG,
+			ciphertext, ciphertext_size,
+			plaintext, (size_t *)plaintext_size,
+			hdr->tag, tag_len);
+	if (res != TEE_SUCCESS)
+		goto exit;
+
+	crypto_ops.authenc.final(ctx, ENC_FS_AUTH_ENC_ALG);
+
+exit:
+	free(ctx);
+	return res;
+}
+
+size_t tee_enc_fs_get_file_header_size(void)
+{
+	return sizeof(struct enc_fs_file_header);
+}
+
+TEE_Result tee_enc_fs_file_encryption(uint8_t *plaintext, size_t plaintext_size,
+			uint8_t *data_out, size_t *out_size)
+{
+	TEE_Result res = TEE_SUCCESS;
+	struct enc_fs_file_header file_header;
+	uint8_t fek[ENC_FS_FEK_SIZE];
+
+	res = generate_fek(fek, ENC_FS_FEK_SIZE);
```

Comment:
I think the design is correct, because `file_header.aad.encrypted_key` need to be encrypted by SSK first, then can continue encrypt whole file structure (file header + file content) by FEK using AES-GCM mode. In other word, `file_header.aad.encrypted_key` is used for storing encrypted FEK and can not be used for encrypt whole file structure.